Plugin docs (for plugin developers) (#1030)
Napari/napari: docs/source/plugins/for_plugin_developers.rst

.. _plugins-for-plugin-developers:

.. role:: python(code)
   :language: python

creating a napari plugin
========================

This document explains how to extend napari functionality by writing a plugin
that can be installed with ``pip`` and autodetected by napari.  For in-depth
information on how plugins are used internally in napari, see
:ref:`plugins-for-napari-developers`.

Overview
--------

``napari`` allows developers to extend the functionality of the program as
follows:

1. **Hooks**: We (napari) identify specific places ("*hooks*") in the napari
codebase where we would like external developers to be able to extend
functionality. For example, when a user tries to open a filepath in napari, we
might want to enable plugins to extend the file formats that can be handled.  A

> *hook*, then, is basically just the exact place within napari where we

Reviewer: ```suggestion
*hook*, then, is the place within napari where we
```

😉

"promise" to call functions created by external developers.

2. **Hook Specifications**:  We then create "*hook specifications*", which are

> little more than well-documented function signatures that define the API (or

Reviewer: ```suggestion
well-documented function signatures that define the API (or
```

"contract") that a plugin developer must adhere to when writing their function
that we promise to call somewhere in the napari codebase.

3. **Hook Implementations**: Plugin developers then write functions ("*hook
implementations*") and mark that function as meeting the requirements of a
specific *hook specification* offered by napari (using a decorator as
:ref:`described below <hookimpl-decorator>`).

4. **Plugin discovery**: Using :ref:`one of two methods <hookimpl-decorator>`,
plugins that are installed in the same python environment as napari can make
themselves known to napari. ``napari`` will then scan plugin modules for *hook
implementations* that will then be called at the appropriate time place during

> the execution of ``napari``.

Reviewer: Love the above four points. =)

.. NOTE::
   One of the primary ways that we will extend the functionality of napari over
   time is by identifying new ideas for *hook specifications* that developers
   can implement.  If you have a plugin idea that requires napari to create a
   new hook specification, we'd love to hear about it!  Please think about what
   the signature of your proposed hook specification would look like, and where
   within the napari codebase you'd like your hook implementation to be called,
   and `open a feature request
   <https://github.com/napari/napari/issues/new?template=feature_request.md>`_
   in the napari issue tracker with your proposal.

Let's take the :func:`~napari.plugins.hook_specifications.napari_get_reader`
hook (our primary "reader plugin" hook) as an example.  It is defined as:

.. code-block:: python

   LayerData = Union[Tuple[Any], Tuple[Any, Dict], Tuple[Any, Dict, str]]
   ReaderFunction = Callable[[str], List[LayerData]]

   @napari_hook_specification(firstresult=True)
   def napari_get_reader(
       path: Union[str, List[str]]
   ) -> Optional[ReaderFunction]:
       ...

Note that it takes a ``str`` or a ``list`` of ``str`` and either returns

> ``None`` or a function.  From the docstring of the hook specification, we see

Reviewer: Can you include this docstring here? Or is it too much? Maybe you can link the function here also. (I see the link above the code block, but I would expect "docstring" to be clickable here. =)

Author: I'll make `docstring` a clickable link.  rather not copy the full thing here

that the implementation should return ``None`` if the path is of an
unrecognized format, otherwise it should return a ``ReaderFunction``, which is
a function that takes a ``str`` (the filepath to read) and returns a ``list``
of ``LayerData``, where ``LayerData`` is any one of ``(data,)``, ``(data,
meta)``, or ``(data, meta, layer_type)``.

That seems like a bit of a mouthful!  But it's a precise (though flexible)
contract that we can follow, and know that napari will handle the rest.

Step 2: Write your hook implementation
--------------------------------------

Once you have identified the :ref:`hook specification <hook-specifications>`
that you want to implement, you have to create a *hook implementation*: a
function that accepts the arguments specified by the hook specification
signature and returns a value with the expected return type.

Here's an example hook implementation for
:func:`~napari.plugins.hook_specifications.napari_get_reader` that enables
napari to open an imaginary ``.ext`` filetype.

.. code-block:: python

   from pluggy import HookimplMarker

>    # we'll get to this line and the decorator below in just a minute

Reviewer: maybe specify where this is explained?

Reviewer: And, seeing that the code requires explanation, I wonder whether we should show importing `napari_hook_implementation` here, which is more self-explanatory, even if only on a superficial level. Then, later where it is explained, you can say, "and, bonus, you don't actually need to import napari to create a plugin! tada!" and explain the fancier pluggy version. What do you think? (Also @sofroniewn)

Author: sure.  I was a little conflicted... since I'd really rather people not default to depending on napari for that line alone.  But I agree it reads better... and until a plugin like this can be used by something other than napari, not depending on napari is of minor benefit.

   napari_hook_implementation = HookimplMarker("napari")

   @napari_hook_implementation
   def napari_get_reader(path):
      # remember, path can be a list, so we check it's type first...
      # (this example plugin doesn't handle lists)
      if isinstance(path, str) and path.endswith(".ext"):
         # If we recognize the format, we return the actual reader function
         return my_reader
      # otherwise we return None.
      return None

   def my_reader(path):
      with open(path, 'rb') as file:
         array = convert_bytes_to_numpy(data)
      # return it as a list of LayerData

>       return [(array,)]

Reviewer: I would move this definition above the `napari_get_reader` call and rename it `ext_file_reader`. To quote @stefanv from long ago, it's valid for the Python parser to define the function under the place where it's called, but it breaks my brain's parser. =P (He told *me* when I'd done the exact same thing. 😂)

Reviewer: Also, I don't love that `convert_bytes_to_numpy` is undefined. Maybe just do a real plugin here for .npy files using np.load?

Author: will switch to a numpy reader plugin

.. note::

  The seemingly excessive ``list``-of-``tuples`` return type here allows
  plugins the flexibility of returning multiple layers, with optional

>   layer-construction arguments.

Reviewer: I'd remove this comment as well.

Reviewer: (save it for when the usage of that form is documented.)

.. code-block:: python

   from pluggy import HookimplMarker

   napari_hook_implementation = HookimplMarker("napari")

Currently (as of March, 2020), the only way that napari knows *which* hook
specification your implementation matches is by looking at the *name* of your
function.  So in the example above, it was critical that our hook
implementation was literally named ``napar_get_reader``:

.. code-block:: python

   @napari_hook_implementation
   def napari_get_reader(path: str):
      ...

However, `a pull request has been merged at pluggy
<https://github.com/pytest-dev/pluggy/pull/251>`_ that will enable you to mark
*any* function as satisfying a napari hook specification (regardless of the
function's name) using the following syntax:

.. code-block:: python

   @napari_hook_implementation(specname="napari_get_reader")
   def whatever_name_you_want(path: str):
      ...

(Monitor the `pluggy changelog
<https://github.com/pytest-dev/pluggy/blob/master/CHANGELOG.rst>`_ for release
of PR #251.)

.. _plugin-discover:

Step 3: Make your plugin discoverable
-------------------------------------

Packages and modules installed in the same environment as ``napari`` may make

> themselves "discoverable" to napari using one of two conventions:

Reviewer: maybe link here to the document you showed us, I think from PyPA (?), when we were first discussing this?

^^^^^^^^^^^^^^^^^^^^^^

By providing an ``entry_points`` argument with the key ``napari.plugin`` to
``setup()`` in ``setup.py``, plugins can register themselves for discovery
(even if their names do not begin with "``napari_``").

For example if you have a package named ``mypackage`` with a submodule
``napari_plugin`` where you have decorated one or more napari hook
implementations, then if you include in ``setup.py``:

.. code-block:: python

   # setup.py

   setup(
      ...
      entry_points={'napari.plugin': 'plugin_name = mypackage.napari_plugin'},
      ...
   )

... then napari will search the ``mypackage.napari_plugin`` module for
functions decorated with the ``HookimplMarker("napari")`` decorator and
register them the plugin name ``"plugin_name"``.

One benefit of using this approach is that if you already have an existing
pip-installable package, you can extend support for ``napari`` without having
to rename your package, simply by identifying the module in your package that
has the hook implementations.

A user would then be able to use ``napari``, extended with your package's
functionality by simply installing your package along with napari:

.. code:: bash

>    pip install napari mypackage

Reviewer: I feel like this needs a follow-up line:
```suggestion
   pip install napari mypackage

which is 🎵 ah-maziiing! 🎵 
```

😀

Author: lol.  serious??  I'll do it!

Reviewer 2: umm ..... not so sure about this one ;-)

Reviewer 2: but it is amazing :-)

Reviewer 2: maybe just "which is amazing!"

Author: while I am equally enthused, I think I'll leave it out ;)

Cookiecutter template
---------------------

To quickly generate a new napari plugin project, you may wish to use the
`cookiecutter-napari-plugin
<https://github.com/napari/cookiecutter-napari-plugin>`_ template.  This uses
the `cookiecutter <https://github.com/cookiecutter/cookiecutter>`_ command line
utility, which will ask you a few questions about your project and get you
started with a ready-to-go package layout where you can begin implementing your
plugin.

Install cookiecutter and use the template as follows:

.. code-block:: bash

   $ pip install cookiecutter

>    $ cookiecutter https://github.com/napari/cookiecutter-napari-plugin

Reviewer: You should decide whether you want $ prompts in front of bash code, or not (see above). I think I prefer not, for copy-paste convenience. Ironically, you'd never want to copy the bash code above this one (pip install napari mypackage), but you *would* want to copy the above. =)

Author: no.  I'd prefer to be inconsistent thank you. 😂 ... removed the `$`

Example Plugin
--------------

For a small working plugin example, see the `napari-dv
<https://github.com/tlambert03/napari-dv>`_ plugin, which allows ``napari`` to

> read the ``.dv`` image file format.

Reviewer: Is there a link for this format? Seems very mysterious otherwise... =)

Maybe also link to aicsimageio for a more elaborate example? =)

Author: added a link and also highlight aicsimageio.

Help
----

If you run into trouble creating your plugin, don't hesitate to reach out for
help in the `napari issue tracker

> <https://github.com/napari/napari/issues/new/choose>`_.

Reviewer: Is that the right place? Or maybe image.sc?

Author: image.sc def.  added.  but also suggest to open an issue with bug/feature requests.

.. _plugins-for-plugin-developers:

.. role:: python(code)
   :language: python

creating a napari plugin
========================

This document explains how to extend napari functionality by writing a plugin
that can be installed with ``pip`` and autodetected by napari.  For in-depth
information on how plugins are used internally in napari, see
:ref:`plugins-for-napari-developers`.

> Overview

Reviewer 2: Do you want to mention the existence of the cookiecutter up here with a link below?

Author: added

--------

``napari`` allows developers to extend the functionality of the program as
follows:

1. **Hooks**: We (napari) identify specific places ("*hooks*") in the napari
codebase where we would like external developers to be able to extend
functionality. For example, when a user tries to open a filepath in napari, we
might want to enable plugins to extend the file formats that can be handled.  A
*hook*, then, is the place within napari where we
"promise" to call functions created by external developers.

2. **Hook Specifications**:  We then create "*hook specifications*", which are
well-documented function signatures that define the API (or
"contract") that a plugin developer must adhere to when writing their function
that we promise to call somewhere in the napari codebase.

3. **Hook Implementations**: Plugin developers then write functions ("*hook
implementations*") and mark that function as meeting the requirements of a
specific *hook specification* offered by napari (using a decorator as
:ref:`described below <hookimpl-decorator>`).

> 4. **Plugin discovery**: Using :ref:`one of two methods <hookimpl-decorator>`,

Reviewer 2: You mention "one of two methods" here, but don't say which they are, could you maybe say something and then have the link, otherwise I find myself wanting more information here

Reviewer 2: Maybe just mention a "naming convention" or "package metadata"

Author: added

   in the napari issue tracker with your proposal.

Let's take the :func:`~napari.plugins.hook_specifications.napari_get_reader`
hook (our primary "reader plugin" hook) as an example.  It is defined as:

.. code-block:: python

   LayerData = Union[Tuple[Any], Tuple[Any, Dict], Tuple[Any, Dict, str]]
   ReaderFunction = Callable[[str], List[LayerData]]

   @napari_hook_specification(firstresult=True)
   def napari_get_reader(
       path: Union[str, List[str]]
   ) -> Optional[ReaderFunction]:
       ...

Note that it takes a ``str`` or a ``list`` of ``str`` and either returns
``None`` or a function.  From the :func:`docstring
<napari.plugins.hook_specifications.napari_get_reader>` of the hook
specification, we see that the implementation should return ``None`` if the
path is of an unrecognized format, otherwise it should return a
``ReaderFunction``, which is a function that takes a ``str`` (the filepath to
read) and returns a ``list`` of ``LayerData``, where ``LayerData`` is any one
of ``(data,)``, ``(data, meta)``, or ``(data, meta, layer_type)``.

That seems like a bit of a mouthful!  But it's a precise (though flexible)

> contract that we can follow, and know that napari will handle the rest.

Reviewer 2: ```suggestion
contract that you can follow, and know that napari will handle the rest.
```

functionality by simply installing your package along with napari:

.. code:: bash

   pip install napari mypackage

Cookiecutter template
---------------------

To quickly generate a new napari plugin project, you may wish to use the
`cookiecutter-napari-plugin
<https://github.com/napari/cookiecutter-napari-plugin>`_ template.  This uses
the `cookiecutter <https://github.com/cookiecutter/cookiecutter>`_ command line
utility, which will ask you a few questions about your project and get you
started with a ready-to-go package layout where you can begin implementing your
plugin.

Install cookiecutter and use the template as follows:

.. code-block:: bash

   pip install cookiecutter
   cookiecutter https://github.com/napari/cookiecutter-napari-plugin

Example Plugins
---------------

For a minimal working plugin example, see the `napari-dv
<https://github.com/tlambert03/napari-dv>`_ plugin, which allows ``napari`` to
read the `Priism/MRC/Deltavision image file format
<https://github.com/tlambert03/mrc>`_.

For a more thorough plugin see `napari-aicsimageio
<https://github.com/AllenCellModeling/napari-aicsimageio>`_, one of the first

> third-party plugins developed for napari.  This plugin takes advantage of

Reviewer 2: ```suggestion
community plugins developed for napari.  This plugin takes advantage of
```

I prefer community over third-party, curious what others think. Feels more inclusive

Author: Yeah, I started there actually... I don't know why but somehow it felt like it was "spreading credit" for Jackson's work lol... but I also prefer the word community
